### PR TITLE
Improve error message for SAXExceptions on deployment

### LINKED ISSUE #5490: ModelParseException: SAXException while parsing input stream
**Describe the bug**
Observed as error in production logs.

**Expected behavior**
- SAXParisng exceptions are not logged as error

**Log/Stacktrace**
https://console.cloud.google.com/errors/CK7AzeKf3rSTYA?service=zeebe&time=P7D&project=camunda-cloud-240911
https://console.cloud.google.com/errors/COn7jpDXj4qlPQ?service=zeebe&time=P7D&project=camunda-cloud-240911

**Environment:**
Camunda Cloud / Prod / 0.24.2, 0.24.3

### LINKED ISSUE #8026: Improve error message for SAXExceptions on deployment
**Is your feature request related to a problem? Please describe.**
We had a support issue in which the customer had a malformed XML. They were unable to deploy it and the error given in console was not helpful:

image

I had to look at the logs to figure out what the problem was.

PS: I have not checked whether more information would be available if they had used the Zeebe Client to deploy the model. At least in console, no further information was available)

**Describe the solution you'd like**
Error message for SAXException should contain enough detail for users to figure out what is wrong with the model

**Additional context**
File to provoke the error:
[unhelpful_error.bpmn.zip](https://github.com/camunda-cloud/zeebe/files/7371703/unhelpful_error.bpmn.zip)

### PR DESCRIPTION
**Pull request by skayliu**
## Description

Error message for SAXException should contain enough detail for users to figure out what is wrong with the model.

e.g.
before:
org.camunda.bpm.model.xml.ModelParseException: SAXException while parsing input stream 

after:
org.xml.sax.SAXException: Error: URI=null Line=33: cvc-complex-type.3.2.2: Attribute 'stroke' is not allowed to appear in element 'bpmndi:BPMNPlane'.

## Related issues

closes #8026
closes #5490

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by skayliu** [comment]
Hi, @saig0, @korthout . Please take a moment to review this fix when you have time. Thank you.

**Comment by saig0** [comment]
@skayliu thank you for your contribution. :tada: We'll have a look at your PR in the next few days. :eyes:

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/BpmnResourceTransformer.java [hunk 1]
```diff
@@ -60,24 +60,35 @@ public BpmnResourceTransformer(
...
+      }
+    } catch (final Exception e) {
+      final var cause = e.getCause();
+      final String failureMessage;
+      if (cause instanceof SAXException) {
+        failureMessage =
+            String.format("'%s': SAXException %s", resource.getResourceName(), cause.getMessage());
+      } else {
+        failureMessage = String.format("'%s': %s", resource.getResourceName(), e.getMessage());
+      }
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/BpmnResourceTransformer.java:91]: ❌ I think this too much of a catch-all. I would propose that instead we isolate the part which can throw the exception. This is already isolated as it's thrown by the `readProcessDefinition` method. 

We can change the signature of this method and have it return an `Either<Failure, BpmnModelInstance>`. Inside of the method we need to catch a `ModelParseException` and if we do we return an `Either.left` with the failure containing nice message for users. If no exception is thrown we will return an `Either.right` with the model instance.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Hi @skayliu! Thanks for your contribution 🙇 

I have a comment regarding the approach you took to fix this. Please have a look at it and adjust accordingly. Let me know if you have any questions!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/BpmnResourceTransformer.java (hunk 1 above)
Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/BpmnResourceTransformer.java]: Done.

Hi @remcowesterhoud. Please check this out. :bow:

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java [hunk 2]
```diff
@@ -220,4 +211,43 @@ public void shouldRejectDeploymentIfResourceIsTooLarge() {
...
+     "Expected to deploy new resources, but encountered the following errors:
+     '/processes/saxexception-error-8026.bpmn': SAXException Error: URI=null Line=33:
+     cvc-complex-type.3.2.2: 元素 'bpmndi:BPMNPlane' 中不允许出现属性 'stroke'。"
+
+     In English
+     "Expected to deploy new resources, but encountered the following errors:
+     '/processes/saxexception-error-8026.bpmn': SAXException Error: URI=null Line=33:
+     cvc-complex-type.3.2.2: Attribute 'stroke' is not allowed to appear in element
+     'bpmndi:BPMNPlane'.
+    */
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java:245]: 🔧 After the previous changes this comment is wrong. I think we can just remove this comment entirely.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java [hunk 3]
```diff
@@ -220,4 +211,43 @@ public void shouldRejectDeploymentIfResourceIsTooLarge() {
...
+
+    assertThat(deploymentRejection.getRejectionReason())
+        .describedAs("rejection should contain enough detail rejection reason")
+        .contains("cvc-complex-type.3.2.2")
+        .contains("stroke")
+        .contains("bpmndi:BPMNPlane");
+  }
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java:252]: 🔧 Let's also verify that it adds the name of the bpmn file

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @lzgabel! I have 2 minor comments on that test. If these are fixed I can merge it 🚀

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java (hunk 3 above)
Comment (lzgabel) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java]: Done.

Hi @remcowesterhoud. Please check this out. ❤️

**Review by remcowesterhoud — APPROVED** [review]
⚡ fast, thanks both 🙇 

bors merge

**Comment by remcowesterhoud** [comment]
Yeah, please give that a try and I will retry the merge

**Comment by lzgabel** [comment]
Done. 

@remcowesterhoud. Please try again. 🙇

**Comment by remcowesterhoud** [comment]
bors merge

**Pull request merged**